lumigator/backend/backend/tracking/mlflow.py

```diff
@@ -14,66 +28,354 @@ class MLflowTrackingClient(TrackingClient):
+                                )
+                        # merge the results into the compiled results
+                        compiled_results.update(job_results)
+            with s3.open(f"{settings.S3_BUCKET}/{workflow_id}/compiled.json", "w") as f:
+                f.write(json.dumps(compiled_results))
+            # Generate presigned download URL for the object
+        s3_client = boto3.client("s3", endpoint_url=settings.S3_ENDPOINT_URL)
+        download_url = s3_client.generate_presigned_url(
+            "get_object",
+            Params={
+                "Bucket": settings.S3_BUCKET,
+                "Key": f"{workflow_id}/compiled.json",
+            },
+            ExpiresIn=settings.S3_URL_EXPIRATION,
+        )
+        workflow_details.artifacts_download_url = download_url
+        return workflow_details
 
-    def list_experiments(self) -> list:
-        raise NotImplementedError
+    def update_workflow_status(self, workflow_id: str, status: WorkflowStatus) -> None:
+        """Update the status of a workflow."""
+        self._client.set_tag(workflow_id, "status", status)
 
-    def get_workflow(self, workflow_id: str) -> dict:
-        raise NotImplementedError
+    def _get_ray_job_logs(self, ray_job_id: str):
+        """Get the logs for a Ray job."""
+        resp = requests.get(
+            urljoin(settings.RAY_JOBS_URL, f"{ray_job_id}/logs"), timeout=5
+        )  # 5 sec
 
-    def update_workflow(self, workflow_id: str, new_data: dict) -> None:
-        raise NotImplementedError
+        if resp.status_code == HTTPStatus.NOT_FOUND:
+            loguru.logger.error(
+                f"Upstream job logs not found: {resp.status_code}, error: {resp.text or ''}"
+            )
+            raise HTTPException(
+                status_code=HTTPStatus.NOT_FOUND,
+                detail=f"Job logs for ID: {ray_job_id} not found",
+            )
+        elif resp.status_code != HTTPStatus.OK:
+            loguru.logger.error(
+                f"Unexpected status code getting job logs: {resp.status_code}, \
+                    error: {resp.text or ''}"
+            )
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
+                detail=f"Unexpected error getting job logs for ID: {ray_job_id}",
+            )
 
-    def delete_workflow(self, workflow_id: str) -> None:
-        raise NotImplementedError
+        try:
+            metadata = json.loads(resp.text)
+            return JobLogsResponse(**metadata)
+        except json.JSONDecodeError as e:
+            loguru.logger.error(f"JSON decode error: {e}")
+            loguru.logger.error(f"Response text: {resp.text}")
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR, detail="Invalid JSON response"
+            ) from e
+
+    def get_workflow_logs(self, workflow_id: str) -> JobLogsResponse:
```

Comment:
The caveat here is that it will only retrieve the logs of jobs that have finished: so this route won't give you any sort of live stream of the logs as it works now. I'm not sure if this is a dealbreaker? Or if it's ok since IIRC there were some other kinks in the logging to work out anyways? Theres more work to do with the logging and we are tracking that via https://github.com/mozilla-ai/lumigator/issues/588

Reply:
@aittalam  I'd be interested in your feedback for this one: is this an ok starting point for log access? If we need the lives to be a log stream I can make that work, but it will take some refactoring (and further increase the size of this PR 😆 )